docs/src/keymap-upgrade/keycodes.ts

```diff
@@ -0,0 +1,166 @@
+  RABT: "GT",
+  COLN: "COLON",
+  KSPC: null,
+  ATSN: "AT",
+  BANG: "EXCL",
+  LCTL: "LCTRL",
+  LSFT: "LSHIFT",
+  RCTL: "RCTRL",
+  RSFT: "RSHIFT",
+  M_NEXT: "C_NEXT",
+  M_PREV: "C_PREV",
+  M_STOP: "C_STOP",
+  M_EJCT: "C_EJECT",
+  M_PLAY: "C_PP",
+  M_MUTE: "C_MUTE",
+  M_VOLU: "C_VOL_UP",
+  M_VOLD: "C_VOL_DN",
+  GUI: "K_CMENU",
+  MOD_LCTL: "LCTRL",
+  MOD_LSFT: "LSHIFT",
+  MOD_LALT: "LALT",
+  MOD_LGUI: "LGUI",
+  MOD_RCTL: "RCTRL",
+  MOD_RSFT: "RSHIFT",
+  MOD_RALT: "RALT",
+  MOD_RGUI: "RGUI",
+};
+
+/**
+ * Upgrades deprecated key code identifiers.
+ */
+export function upgradeKeycodes(tree: Tree) {
+  const edits: TextEdit[] = [];
+
+  const query = Devicetree.query("(identifier) @name");
+  const matches = query.matches(tree.rootNode);
+
+  for (const { captures } of matches) {
+    const node = findCapture("name", captures);
+
+    // Some of the codes are still valid to use in other properties such as
+    // "mods", so only replace those that are inside a "bindings" array.
+    if (node && isInBindingsArray(node)) {
+      edits.push(...getUpgradeEdits(node, CODES, keycodeReplaceHandler));
+    }
+  }
+
+  return edits;
+}
+
+function keycodeReplaceHandler(node: SyntaxNode, replacement: string | null) {
+  if (replacement) {
+    return [TextEdit.fromNode(node, replacement)];
+  }
+
+  const nodes = findBehaviorNodes(node);
+
+  if (nodes.length === 0) {
+    console.warn(
+      `Found deprecated code "${node.text}" but it is not a parameter to a behavior`
+    );
+    return [TextEdit.fromNode(node, `/* "${node.text}" no longer exists */`)];
+  }
+
+  const oldText = nodes.map((n) => n.text).join(" ");
+  const newText = `&none /* "${oldText}" no longer exists */`;
+
+  const startIndex = nodes[0].startIndex;
+  const endIndex = nodes[nodes.length - 1].endIndex;
+
+  return [new TextEdit(startIndex, endIndex, newText)];
+}
+
+/**
+ * Given a parameter to a keymap behavior, returns a list of nodes beginning
+ * with the behavior and including all parameters.
+ * Returns an empty array if no behavior was found.
+ */
+function findBehaviorNodes(paramNode: SyntaxNode) {
+  // Walk backwards from the given parameter to find the behavior reference.
+  let behavior = paramNode.previousNamedSibling;
+  while (behavior && behavior.type !== "reference") {
+    behavior = behavior.previousNamedSibling;
+  }
+
+  if (!behavior) {
+    return [];
+  }
+
+  // Walk forward from the behavior to collect all its parameters.
+  let nodes = [behavior];
+  let param = behavior.nextNamedSibling;
+  while (param && param.type !== "reference") {
+    nodes.push(param);
+    param = param.nextNamedSibling;
+  }
+
+  return nodes;
+}
+
+/**
+ * Given a identifier, returns whether it is inside a "bindings" property value.
+ */
+function isInBindingsArray(identifier: SyntaxNode) {
+  let node = identifier.parent;
+  while (node) {
+    if (node.type === "property") {
+      return node.childForFieldName("name")?.text === "bindings";
```

Comment:
Might be worth checking for `sensor-bindings` as well, if not a regex/suffix to cover future cases.

Reply:
Good catch. Done.